### PR TITLE
support ca-file parameter in CSI

### LINKED ISSUE #498: Support ca-file option when using csi
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
/kind bug
> /kind feature

**What happened**:
Currently, when creating Openstack instance in csi, ca-file option is not considered. This should be supported 
**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- openstack-cloud-controller-manager version:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by ramineni**
**What this PR does / why we need it**:
currently, openstack client created in csi, doesn't support ca-file parameter . This PR is to update the same.
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #498 #499

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

File: pkg/csi/cinder/openstack/openstack.go [hunk 1]
```diff
@@ -61,6 +64,7 @@ type Config struct {
 		DomainId   string `gcfg:"domain-id"`
 		DomainName string `gcfg:"domain-name"`
 		Region     string
+		CAFile     string `gcfg:"ca-file"`
```
Comment (jichenjc) [inline pkg/csi/cinder/openstack/openstack.go]: just a reminder, later on we can add a section in doc about the CAFile setting..

File: pkg/csi/cinder/openstack/openstack.go [hunk 2]
```diff
@@ -94,33 +98,31 @@ func isNotFound(err error) bool {
...
 }
 
-func GetConfigFromFile(configFilePath string) (gophercloud.AuthOptions, gophercloud.EndpointOpts, error) {
-	// Get config from file
-	var authOpts gophercloud.AuthOptions
+// GetConfigFromFile retrieves config options from file
+func GetConfigFromFile(configFilePath string) (Config, gophercloud.EndpointOpts, error) {
 	var epOpts gophercloud.EndpointOpts
+	var cfg Config
 	config, err := os.Open(configFilePath)
```
Comment (ningning54321) [inline pkg/csi/cinder/openstack/openstack.go]: typo: config->cfg?

**Comment by adisky** [comment]
@darcyllingyan can you test this PR?

**Comment by darcyllingyan** [comment]
@adisky ,
Thanks for the quick response!
The PR has been involved to the master branch? And I just need to get the lastest code from master branch and test it, right?

Thanks
Darcy

**Comment by adisky** [comment]
@darcyllingyan can you join slack
https://kubernetes.slack.com/messages/sig-openstack
we can discuss there

**Comment by darcyllingyan** [comment]
@adisky ,
I have just created a new slack workspace, not sure whether it's OK?
https://join.slack.com/t/kubernetes-kvx3616/shared_invite/enQtNTU4MDIyMDMzMzc4LTA5ZDcxNzk1ZDg0MTNlMjM2MzAxMGZmNmE1MDNiYzM0ZjMxZmVjYjQ1YWJhZGI4MjIzYjY4MzkzOTJlODMwOTE

**Comment by jichenjc** [comment]
@darcyllingyan you need join https://kubernetes.slack.com/messages/C0LSA3T7C/ (sig-openstack)

**Comment by darcyllingyan** [comment]
@jichenjc ,
Thanks for the response!
I didn't have the account like @get-an-invite-at.slack.kubernetes.io email address and in the link :http://slack.kubernetes.io/ shows "Kubernetes Slack invites are currently unavailable", so seems I couldn't login now.

**Comment by adisky** [comment]
@darcyllingyan It seems there is some issue going on for new inivites,
you need to pull this changes and build the ciner csi container image using 
```make image-csi-plugin```, tag this image to latest and run deploy it

**Comment by darcyllingyan** [comment]
@adisky ,
OK, thanks.
I will have a try.

Thanks
Darcy

**Comment by darcyllingyan** [comment]
Hi @adisky ,

>you need to pull this changes and build the ciner csi container image using
`make image-csi-plugin`, tag this image to latest and run deploy it

There are some issues when I build the cinder csi cintainer images.
1. When I use the hack/make image-csi-plugin
 the process is below:
```
docker run -it -v "$PWD":/go/src/k8s.io/cloud-provider-openstack:z -w /go/src/k8s.io/cloud-provider-openstack golang:1.10
root@761f320a0021:/go/src/k8s.io/cloud-provider-openstack# 
root@761f320a0021:/go/src/k8s.io/cloud-provider-openstack# make image-csi-plugin
(99/102) Wrote k8s.io/client-go@kubernetes-1.13.0
(100/102) Wrote k8s.io/api@kubernetes-1.13.0
(101/102) Wrote github.com/docker/docker@a9fbbdc8dd8794b20af358382ab780559bca589d
(102/102) Wrote k8s.io/kubernetes@v1.13.0
CGO_ENABLED=0 GOOS=linux go build \
        -ldflags "-w -s -X 'main.version=4914f388-dirty'" \
        -o cinder-csi-plugin \
        cmd/cinder-csi-plugin/main.go
cp cinder-csi-plugin cluster/images/cinder-csi-plugin
docker build -t k8scloudprovider/cinder-csi-plugin:4914f388-dirty cluster/images/cinder-csi-plugin
make: docker: Command not found
Makefile:233: recipe for target 'image-csi-plugin' failed
make: *** [image-csi-plugin] Error 127

```

In the golang container, running the ```make image-csi-plugin``` couldn't succeed as  the docker, yum, vi, rpm and many packages are not installed, right?
 
2. If I first create the golang environment, use the ```make image-csi-plugin```, following errors are reported:
```
# make image-csi-plugin
dep ensure -v
CGO_ENABLED=0 GOOS=linux go build \
        -ldflags "-w -s -X 'main.version=4914f388-dirty'" \
        -o cinder-csi-plugin \
        cmd/cinder-csi-plugin/main.go
cp cinder-csi-plugin cluster/images/cinder-csi-plugin
docker build -t k8scloudprovider/cinder-csi-plugin:4914f388-dirty cluster/images/cinder-csi-plugin
Sending build context to Docker daemon 28.17 MB
Step 1/6 : FROM centos:7.4.1708
 ---> 295a0b2bd8ea
Step 2/6 : LABEL maintainers "Kubernetes Authors"
 ---> Using cache
 ---> d6b0662dc6d0
Step 3/6 : LABEL description "Cinder CSI Plugin"
 ---> Using cache
 ---> 387122440087
Step 4/6 : RUN yum -y install e4fsprogs
 ---> Running in 677ac4bd9d70

Loaded plugins: fastestmirror, ovl


 One of the configured repositories failed (Unknown),
 and yum doesn't have enough cached data to continue. At this point the only
 safe thing yum can do is fail. There are a few ways to work "fix" this:

     1. Contact the upstream for the repository and get them to fix the problem.

     2. Reconfigure the baseurl/etc. for the repository, to point to a working
        upstream. This is most often useful if you are using a newer
        distribution release than is supported by the repository (and the
        packages for the previous distribution release still work).

     3. Run the command with the repository temporarily disabled
            yum --disablerepo=<repoid> ...

     4. Disable the repository permanently, so yum won't use it by default. Yum
        will then just ignore the repository until you permanently enable it
        again or use --enablerepo for temporary usage:

            yum-config-manager --disable <repoid>
        or
            subscription-manager repos --disable=<repoid>

     5. Configure the failing repository to be skipped, if it is unavailable.
        Note that yum will try to contact the repo. when it runs most commands,
        so will have to try and fail each time (and thus. yum will be be much
        slower). If it is a very temporary problem though, this is often a nice
        compromise:

            yum-config-manager --save --setopt=<repoid>.skip_if_unavailable=true

Cannot find a valid baseurl for repo: base/7/x86_64
Could not retrieve mirrorlist http://mirrorlist.centos.org/?release=7&arch=x86_64&repo=os&infra=container error was
14: curl#7 - "Failed to connect to 2001:4178:5:200::10: Network is unreachable"
The command '/bin/sh -c yum -y install e4fsprogs' returned a non-zero code: 1
make: *** [image-csi-plugin] Error 1
```

I have tried many methods to solve the above problem, and I didn't succeed, do you have any suggestions? Thanks!

**Comment by adisky** [comment]
@darcyllingyan Thanks for trying https://github.com/kubernetes/cloud-provider-openstack/pull/500#issuecomment-466748110, seems you have some restrictions in your env
/lgtm

File: pkg/csi/cinder/openstack/openstack.go (hunk 1 above)
Comment (dims) [reply pkg/csi/cinder/openstack/openstack.go]: Good idea @jichenjc

**Comment by dims** [comment]
/approve

@ramineni can you please update docs in another PR?

**Pull request merged**